### ISSUE TITLE
Connection problems in TCP/IP parallelization

### ISSUE THREAD

**Issue by lguas** [tensorforce/tensorforce#821]
Hi,

I am having some difficulties in using the TCP/IP parallelization of the agent training. 
I am using a custom Gym environment and the serial implementation works with no problem.

I am currently considering 2 server environments. With everything running on the same computer (`host='127.0.0.1'`), I run into the error:
```
Traceback (most recent call last):
  File "/home/luca/anaconda3/envs/tflow/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/home/luca/anaconda3/envs/tflow/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/luca/Documenti/PhD/019-Channel2D_DRL/DRL_Channel2DBubble/src/channel2Dbubble/__main__.py", line 17, in <module>
    args.cmd(**vars(args))
  File "/home/luca/Documenti/PhD/019-Channel2D_DRL/DRL_Channel2DBubble/src/run.py", line 99, in run
    port=conf.runner.port)
  File "/home/luca/Documenti/PhD/019-Channel2D_DRL/tensorforce/tensorforce/execution/runner.py", line 181, in __init__
    assert util.is_equal(x=environment.states(), y=states)
AssertionError
```
Using python debugger I found that there is a problem with the TCP/IP connection:
```
> /home/luca/Documenti/PhD/019-Channel2D_DRL/tensorforce/tensorforce/execution/runner.py(181)__init__()
-> assert util.is_equal(x=environment.states(), y=states)
(Pdb) environment.states()
*** OSError: [Errno 107] Transport endpoint is not connected
```
What I find confusing is that the the first server-client environment couple seems correctly connected, while the second one is failing:
```
> /home/luca/Documenti/PhD/019-Channel2D_DRL/tensorforce/tensorforce/execution/runner.py(168)__init__()
-> states = environment.states()
(Pdb) environment.states()
{'type': 'float', 'shape': [12], 'min_value': None, 'max_value': None}
(Pdb) c
> /home/luca/Documenti/PhD/019-Channel2D_DRL/tensorforce/tensorforce/execution/runner.py(181)__init__()
-> assert util.is_equal(x=environment.states(), y=states)
(Pdb) environment.states()
*** OSError: [Errno 107] Transport endpoint is not connected
```
If I check the port status before the creation of the second environment I get:
```
root@debian:/home/luca# lsof -i -P -n | grep 3500*
python3   23677       luca   16u  IPv4 663967      0t0  TCP *:35001 (LISTEN)
python3   23678       luca   17u  IPv4 663963      0t0  TCP 127.0.0.1:35000->127.0.0.1:42646 (ESTABLISHED)
python3   23729       luca    6u  IPv4 674872      0t0  TCP 127.0.0.1:42646->127.0.0.1:35000 (ESTABLISHED)
```
while after:
```
root@debian:/home/luca# lsof -i -P -n | grep 3500*
python3   23678       luca   17u  IPv4 663963      0t0  TCP 127.0.0.1:35000->127.0.0.1:42646 (ESTABLISHED)
python3   23729       luca    6u  IPv4 674872      0t0  TCP 127.0.0.1:42646->127.0.0.1:35000 (ESTABLISHED)
python3   23729       luca    7u  IPv4 688644      0t0  TCP 127.0.0.1:49570->127.0.0.1:35001 (CLOSE_WAIT)
```
It seems like there is a problem on the server side of the environment, but I was not able to find it out. Any thoughts on this?

Thank you very much!
(Tagging @jerabaul29 as he's collaborating on this project, too)

**Comment by jerabaul29** [comment]
Mmmh, can you post how your script looks like here to get some help? :) There is a wealth of ```*** OSError: [Errno 107] Transport endpoint is not connected``` errors over the internet, maybe more information will be needed to help.

**Comment by jerabaul29** [comment]
Another point: are you able to run the examples here / does this work? :

https://github.com/tensorforce/tensorforce/blob/cdd3c464a3541854b14c53b709372ec7187d9cf4/examples/parallelization.py#L77-L105

And are you following the same kind of use pattern in your code?

**Comment by jerabaul29** [comment]
Looks like this functionality is tested too; could you try to run the test on your machine?

https://github.com/tensorforce/tensorforce/blob/cdd3c464a3541854b14c53b709372ec7187d9cf4/test/test_examples.py#L518-L540

**Comment by AlexKuhnle** [comment]
Agree with @jerabaul29, both checking the parallelization example and posting your script would be helpful. 👍

**Comment by jerabaul29** [comment]
Any update on this @lguas ? :) Any terminal outputs / logs to share? :)

**Comment by lguas** [comment]
I tried to run the parallelization script and it worked correctly. My code is slightly different, as the server environments are called from a separate bash script. This does not affect the functionality, as verified with this [Gist](https://gist.github.com/lguas/281a06d5796e73866b0db6d634a66883)

After updating to the last release, today I found that this behaviour depended on a bug in the custom environment, that provided a 0-dimensional observation state for the environment.
Closing as not related dependent on Tensorforce.
